sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala

```diff
@@ -88,6 +88,10 @@ class ArtifactManager(session: SparkSession) extends Logging {
    */
   protected val sessionArtifactAdded = new AtomicBoolean(false)
 
+  protected val cachedClassLoader = new ThreadLocal[ClassLoader] {
+    override def initialValue(): ClassLoader = null
+  }
+
```

Comment:
I'm not sure I understand why this is a ThreadLocal? Artifacts can be added from multiple threads so wouldn't the addition on one thread cause all the others cached classloaders to be invalidated?

Reply:
> the addition on one thread cause all the others cached classloaders to be invalidated

No, in this case, each thread will have its own classloader, and CodeGen will happen once for each thread (after which will be cached)

Good point. After checking codes about the CodeGen cache I believe it's safe to share classloaders across threads.